nhs/data/filter.py

```diff
@@ -1,26 +1,147 @@
-from typing import List
+import glob
 
 import polars as pl
 
 
+def load_gnaf_files_by_states(
+    gnaf_path: str, states: list[str] = []
+) -> tuple[pl.LazyFrame, pl.LazyFrame]:
+    """
+    Load and filter ADDRESS_DEFAULT_GEOCODE and ADDRESS_DETAIL files for the specified states
+    from the given GNAF directory, and return them as LazyFrames.
+
+    Parameters:
+    - gnaf_path: str, the directory path where the GNAF psv files are stored.
+    - states: list[str], an optional list of state/territory names (e.g., ["WA", "ACT"]).
+              If an empty list is provided, the function will default to including all states.
+
+    Returns:
+    - tuple: (default_geocode_lf, address_detail_lf)
+      - default_geocode_lf: pl.LazyFrame, the merged ADDRESS_DEFAULT_GEOCODE data for the specified states,
+                            with an added "STATE" column.
+      - address_detail_lf: pl.LazyFrame, the merged ADDRESS_DETAIL data for the specified states.
+    """
+    # State codes
+    all_state_codes = ["NSW", "ACT", "VIC", "QLD", "SA", "WA", "TAS", "NT", "OT"]
+
+    # If states is an empty list, use all state codes
+    if not states:
+        states = all_state_codes
+
+    default_geocode_lfs = []
+    address_detail_lfs = []
+
+    for state_name in states:
+        # Find the corresponding ADDRESS_DEFAULT_GEOCODE file
+        geocode_file_pattern = (
+            f"{gnaf_path}/{state_name}_ADDRESS_DEFAULT_GEOCODE_psv.psv"
+        )
+        geocode_files = glob.glob(geocode_file_pattern)
+
+        # Load the found file
+        for geocode_file in geocode_files:
+            lf = pl.scan_csv(
+                geocode_file, separator="|"
+            )  # Read PSV file using '|' separator
+            # Add the state column
+            lf = lf.with_columns(pl.lit(state_name).alias("STATE"))
+            default_geocode_lfs.append(lf)
```

Comment:
also you can simplify this greatly, use `read_spreadsheets` to get a dictionary of lazyframes, then do `{key, [lf.with_columns(state_name).alias(...) for state_name in states] for key, lf in lf_dict.item()}` instead of appending to a list, that reduce this block down to 2 lines